### PR TITLE
MeshBlock Cleanup and Smart Backpointers. Resolves #306.

### LINKED ISSUE #306: Back pointers to meshblocks should be shared pointers.
Since MR #303 makes the meshblock list own shared pointers, we can now take advantage of this with our back pointers, as @AndrewGaspar points out.

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

As discussed in #306 and #303, smart pointers to meshblocks means we can use them elsewhere in the code, for example, with weak back pointers. This Implements that change.

In addition, I substantially clean up the MeshBlock class. I noticed that all three constructors had the same functionality copy-pasted. I therefore removed all but one of them. I also changed the constructors to a factory method. This was necessary because of a subtlety with smart pointers. You cannot create a weak pointer to an object in its own constructor.

## PR Checklist

- [x] Code passes cpplint
- [X] New features are documented.
- [X] Adds a test for any bugs fixed. Adds tests for new features.
- [X] Code is formatted
- [x] Changes are summarized in CHANGELOG.md

### CODE REVIEW

File: src/mesh/amr_loadbalance.cpp
Comment (AndrewGaspar) [inline src/mesh/amr_loadbalance.cpp]: Applies here, but also many other places. If a routine is not going to actually create a new reference to a `shared_ptr` that persists past the scope of the function, then it doesn't really make sense to pass as a `shared_ptr`. Instead, you should really just leave these as `MeshBlock *`. It better expresses intent at the call site (`pmb.get()` to get a `MeshBlock *`) - that you're handing out a temporary borrow rather than creating new shared ownership. The ownership of the object is still asserted at the call site since you have a shared_ptr.

Additionally, it's a bad pattern for performance (doesn't apply in this specific case because it's a big function) - passing a `shared_ptr` by value results in an atomic reference AND de-reference, which can overwhelm the cost of small functions

File: example/advection/advection_driver.cpp
Comment (AndrewGaspar) [inline example/advection/advection_driver.cpp]: Just a note: `lock` will definitely help in case somebody introduces a bug that makes the MeshBlock get destroyed before the containers it owns are destroyed, as it will definitely result in a segmentation fault, as opposed to trying to reference a heap address that's potentially still mapped in user memory space. Keep in mind that it will still be a segmentation fault, not something more informative.

Comment (AndrewGaspar) [reply example/advection/advection_driver.cpp]: If you're going to change all the places `pmy_block` is used anyway, I'd prefer you actually add accessor functions like `std::shared_ptr<MeshBlock> Block() const { return pmy_block.lock(); }` and make `pmy_block` private.

Comment (AndrewGaspar) [reply example/advection/advection_driver.cpp]: You might even put a null check in there and make a loud error like "Parthenon internal error: Block object destroyed too early!"

File: src/mesh/mesh.hpp
Comment (AndrewGaspar) [inline src/mesh/mesh.hpp]: Can this be private?

File: src/mesh/mesh.hpp
Comment (AndrewGaspar) [inline src/mesh/mesh.hpp]: Hm... I'm not sure this constructor is necessary. Doesn't really reduce LOC at all - seems to increase it, actually. If there was some optimization we could take because we're setting the neighbors at the same time as initialization, but that doesn't seem to be the case.

**Review by AndrewGaspar — CHANGES_REQUESTED** [review]

File: src/mesh/mesh.hpp
Comment (Yurlungur) [reply src/mesh/mesh.hpp]: If we always use the factory functions, yes. But I figured there might be cases where a non-factory call will be made.

File: src/mesh/mesh.hpp
Comment (Yurlungur) [reply src/mesh/mesh.hpp]: It's just syntactic sugar. I'm not married to it.

File: src/mesh/amr_loadbalance.cpp
Comment (Yurlungur) [reply src/mesh/amr_loadbalance.cpp]: Yeah ok. This makes sense. I'll undo those changes.

File: src/mesh/mesh.hpp
Comment (AndrewGaspar) [reply src/mesh/mesh.hpp]: In general for API design, I prefer to limit new APIs by default, and only introduce them when need has arisen. The issue with adding APIs because you anticipate their usage is that they tend not to be built-to-fit.

File: src/mesh/mesh.hpp
Comment (AndrewGaspar) [reply src/mesh/mesh.hpp]: Yeah, I'd say axe this and call `SetNeighbors` at the call site.

File: example/advection/advection_driver.cpp
Comment (Yurlungur) [reply example/advection/advection_driver.cpp]: Resolved.

File: src/mesh/amr_loadbalance.cpp
Comment (Yurlungur) [reply src/mesh/amr_loadbalance.cpp]: Resolved.

File: src/mesh/mesh.hpp
Comment (Yurlungur) [reply src/mesh/mesh.hpp]: Yeah ok. I suspect something like this may need to be public eventually... or more factory functions may need to be created. But for now I'll make it private.

File: src/mesh/mesh.hpp
Comment (Yurlungur) [reply src/mesh/mesh.hpp]: Resolved.

**Comment by Yurlungur** [comment]
@AndrewGaspar I believe all your requested changes have been implemented.

File: src/mesh/mesh.hpp
Comment (AndrewGaspar) [inline src/mesh/mesh.hpp:261]: Funny this wasn't caught until now.

**Review by AndrewGaspar — APPROVED** [review]
`GetBlockPointer` is a mouthful, but I'll accept it.

**Comment by Yurlungur** [comment]
@pgrete @forrestglines can I get one more approval from the Athena team?

**Comment by AndrewGaspar** [comment]
Yo @github just go ahead and merge this when the CI passes - could you do that? No?

**Comment by Yurlungur** [comment]
> `GetBlockPointer` is a mouthful, but I'll accept it.

Yeah. I know. `Block` felt too vague, though.

File: src/interface/container.hpp
Comment (pgrete) [inline src/interface/container.hpp]: [nitpicking]
Should be potentially call this SetBlockPointer (so that it matches GetBlockPointer).
As far as I see this is only called like 5 times in the codebase so the extra characters wouldn't clutter too much
[/nitpicking]

File: src/utils/buffer_utils.hpp
Comment (pgrete) [inline src/utils/buffer_utils.hpp]:
```suggestion
```

File: src/utils/buffer_utils.cpp
Comment (pgrete) [inline src/utils/buffer_utils.cpp]:
```suggestion
```

File: src/mesh/meshblock.cpp
Comment (pgrete) [inline src/mesh/meshblock.cpp:159]: Maybe I'm missing something, but where does this happen now?

Comment (Yurlungur) [reply src/mesh/meshblock.cpp:159]: It happens [here](https://github.com/lanl/parthenon/blob/a004fa7ac8f5f7964a62b0f68d6bebb0344105d6/src/mesh/meshblock.cpp#L126). There were three `MeshBlock` constructors, which all (as far as I can tell) were identical but with slightly different signatures and duplicated code. So I combined them.

File: src/interface/container.hpp
Comment (Yurlungur) [reply src/interface/container.hpp]: Done.

File: src/mesh/meshblock.cpp
Comment (pgrete) [reply src/mesh/meshblock.cpp:159]: Silly me... I just looked at the diff and didn't look at the whole file. Thanks for clarifying.

**Review by pgrete — APPROVED** [review]
Looks good and the cleanup of the constructors was well overdue.
Given the extent of the changes I'll trigger the full test now.
Once that passes this can get merged from my point of view.

**Comment by Yurlungur** [comment]
Pulling the trigger.

**Pull request merged**
